fl4health/utils/metrics.py

```diff
@@ -133,140 +260,51 @@ def __call__(self, logits: torch.Tensor, target: torch.Tensor) -> Scalar:
-    the metrics average is accurate. The compute method is used to return a dictionairy of metrics along with their
-    current values.
-    """
-
-    def __init__(self, metrics: Sequence[Metric], name: str = "") -> None:
-        super().__init__(metrics, name)
-        self.metric_values_history: List[List[Scalar]] = [[] for _ in range(len(self.metrics))]
-        self.counts: List[int] = []
-
-    def update(self, input: torch.Tensor, target: torch.Tensor) -> None:
+class MetricManager:
+    def __init__(self, metrics: Sequence[Metric], metric_mngr_name: str) -> None:
         """
-        Evaluate metrics and store results.
-        """
-        metric_values: List[Scalar] = [metric(input, target) for metric in self.metrics]
-        self.counts.append(target.size(0))
-
-        for i, metric_value in enumerate(metric_values):
-            self.metric_values_history[i].append(metric_value)
+        Class to manage a set of metrics associated to a given prediction type.
 
-    def compute(self) -> Dict[str, Scalar]:
-        """
-        Returns average of each metrics given its historical values and counts
+        Args:
+            metrics (Sequence[Metric]): List of metric to evaluate predictions on.
+            metric_mngr_name (str): Name of the metric manager (ie train, val, test)
         """
-        total_count = sum(self.counts)
-        weights: List[float] = [count / total_count for count in self.counts]
-
-        metric_value_averages = []
-        for metric_values in self.metric_values_history:
-            avg = sum([weight * float(val) for weight, val in zip(weights, metric_values)])
-            metric_value_averages.append(avg)
-
-        results: Dict[str, Scalar] = {
-            f"{self.name}_{str(metric)}".lstrip("_"): avg for metric, avg in zip(self.metrics, metric_value_averages)
-        }
-
-        return results
-
-    def clear(self) -> None:
-        self.metric_values_history = [[] for _ in range(len(self.metrics))]
-        self.counts = []
+        self.og_metrics = metrics
+        self.metric_mngr_name = metric_mngr_name
+        self.metrics_per_prediction_type: Dict[str, Sequence[Metric]] = {}
 
+    def update(self, preds: Dict[str, torch.Tensor], target: torch.Tensor) -> None:
+        """
+        Updates (or creates then updates) a list of metrics for each prediction type.
 
-class MetricMeterManager:
-    """
-    Class to manage one or metric meters.
-    """
+        Args:
+            preds (Dict[str, torch.Tensor]): A dictionairy of
+        """
+        if len(self.metrics_per_prediction_type) == 0:
+            self.metrics_per_prediction_type = {key: copy.deepcopy(self.og_metrics) for key in preds.keys()}
 
-    def __init__(self, key_to_meter_map: Dict[str, MetricMeter]):
-        self.key_to_meter_map = key_to_meter_map
+        for pred, mtrcs in zip(preds.values(), self.metrics_per_prediction_type.values()):
```

Comment:
I'm a bit wary of doing the zip here, as it assumes that the keys of the dictionaries are ordered in the same way and are the same length and won't fail if that is not the case. Maybe we do something like
```
assert len(preds) == len(self.metrics_per_prediction_type)
for prediction_key, pred in prediction_keys.items():
    metrics_for_prediction_type = self.metrics_per_prediction_type[prediction_key]
    for metric_for_prediction_type in metrics_for_prediction_type:
        metric_for_prediction_type.update(pred, target)
```

Reply:
Yeah I agree this is a better way to go about it. Just one small issue, the length of the metric_per_prediction_type is going to as long as as the number of prediction types not the as the number of metrics. Thus I instead assert that the list of metrics at a given key is long as preds.